### PR TITLE
Fixed #31757 -- Adjusted system check for SECRET_KEY to warn about autogenerated default keys.

### PR DESCRIPTION
**Pull request by kosc**
Hello!

It's my first attempt of contribution into Django project. If there is some mistakes please point me. I did it by guide the contribution guide. Can add something if needed.

### CODE REVIEW

File: django/core/checks/security/base.py [hunk 1]
```diff
@@ -68,8 +68,9 @@
 
 W009 = Warning(
     "Your SECRET_KEY has less than %(min_length)s characters or less than "
-    "%(min_unique_chars)s unique characters. Please generate a long and random "
-    "SECRET_KEY, otherwise many of Django's security-critical features will be "
+    "%(min_unique_chars)s unique characters, or it's django-generated secret key starting with dj::insecure."
```
Comment (rfleschenberg) [inline django/core/checks/security/base.py:71]: `it's django-generated` --> `it's a Django-generated`

File: django/core/checks/security/base.py [hunk 2]
```diff
@@ -68,8 +68,9 @@
...
+    "Please generate a long and random SECRET_KEY, "
```
Comment (rfleschenberg) [inline django/core/checks/security/base.py:72]: Maybe `"Please generate a sufficiently long and random SECRET_KEY"`? Or should we maybe repeat `min_unique_chars` here to be clear?

File: django/core/management/utils.py [hunk 3]
```diff
@@ -79,7 +79,7 @@ def get_random_secret_key():
     Return a 50 character random string usable as a SECRET_KEY setting value.
     """
     chars = 'abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+)'
-    return get_random_string(50, chars)
+    return 'dj::insecure' + get_random_string(38, chars)
```
Comment (rfleschenberg) [inline django/core/management/utils.py:82]: Maybe ``return 'dj::insecure-' + get_random_string(38, chars)`` to improve the readability of the value a little?

Also, I think we should probably keep the 50 random characters so we don't lose randomness in comparison to the current behavior.

Comment (kosc) [reply django/core/management/utils.py:82]: Oh, it's ok. I've though about - sign but decided to do exactly as in ticket. I agree about keep 50 signs, so I will edit code and the tests checking default SECRET_KEY's length very soon. Thanks a lot for your review!

File: django/core/checks/security/base.py (hunk 1 above)
Comment (kosc) [reply django/core/checks/security/base.py]: Yes, I sometimes failing with English, thank you!

File: django/core/checks/security/base.py (hunk 2 above)
Comment (kosc) [reply django/core/checks/security/base.py]: I don't think we should repeat information from previous line, but agree with your version of sentence.

File: tests/user_commands/tests.py [hunk 4]
```diff
@@ -331,9 +331,9 @@ def test_no_existent_external_program(self):
 
     def test_get_random_secret_key(self):
         key = get_random_secret_key()
-        self.assertEqual(len(key), 50)
+        self.assertEqual(len(key), 63)
```
Comment (rfleschenberg) [inline tests/user_commands/tests.py:334]: Maybe add a comment here explaining why it's `50 + 13`. Overall it looks good to me!

**Comment by adamchainz** [comment]
Can we change the prefix to `django-insecure-` ? I've never enjoyed the shortening of django to 'dj' (it's quite short already, and the d is silent) and I don't think it's been used anywhere officially before. Also using colons seems like a bad idea since they are hard to copy into some config files such as YAML. Users might want to move their secret key to e.g. docker-compose environment variables during development, without changing them from the autogenerated default.

I'd also suggest just `insecure-` as the prefix - not sure the `django` in the name really shows that it's autogenerated.

**Comment by kosc** [comment]
> I'd also suggest just `insecure-` as the prefix - not sure the `django` in the name really shows that it's autogenerated.

Hello! Thank you for reply.
Looks good to me, but who should make a final decision?

**Comment by apollo13** [comment]
> I'd also suggest just insecure- as the prefix - not sure the django in the name really shows that it's autogenerated.

The idea was to provide a pattern that one can grep for easily without (many) false positives. But I am open to ideas :)

**Comment by kosc** [comment]
Maybe 'insecure-prefix' or even 'prefix-insecure' in case if we will have another prefixes for another settings?

**Comment by adamchainz** [comment]
Including "prefix" in the prefix is a bit self descriptive for my preferences.

Perhaps `insecure-default-`?

**Comment by rfleschenberg** [comment]
I agree with @apollo13 that using something like `django-insecure` is good for greppability.

File: django/core/checks/security/base.py [hunk 5]
```diff
@@ -68,8 +68,9 @@
 
 W009 = Warning(
     "Your SECRET_KEY has less than %(min_length)s characters or less than "
-    "%(min_unique_chars)s unique characters. Please generate a long and random "
-    "SECRET_KEY, otherwise many of Django's security-critical features will be "
+    "%(min_unique_chars)s unique characters, or it's a Django-generated secret key starting with insecure-default."
+    "Please generate a sufficiently long and random SECRET_KEY, "
+    "otherwise many of Django's security-critical features will be "
```
Comment (carltongibson) [inline django/core/checks/security/base.py:73]: This would need updating in `checks.txt`

File: django/core/management/utils.py [hunk 6]
```diff
@@ -76,10 +76,10 @@ def find_command(cmd, path=None, pathext=None):
 
 def get_random_secret_key():
     """
-    Return a 50 character random string usable as a SECRET_KEY setting value.
+    Return a 50 character random string (with insecure prefix) usable as a SECRET_KEY setting value.
     """
     chars = 'abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+)'
-    return get_random_string(50, chars)
+    return 'insecure-default-' + get_random_string(50, chars)
```
Comment (carltongibson) [inline django/core/management/utils.py:82]: I don't think we can make the change here: it's a perennial SO/blog suggestions to import `get_random_secret_key()` to generate new ones. We'll break people's workflows if we mark such keys as insecure. 

The suggestion from the mailing list thread (I thought) was to just adjust the default settings (template) to prefix the starter key there as insecure.

**Comment by pauloxnet** [comment]
I missed why in the first place the generated secret key is considered insecure. There's a discussion in the ML ? Maybe add it in the PR descriptions.

**Comment by carltongibson** [comment]
@pauloxnet It's because folks commit it to VCS and forget to change from the first one provided.

**Comment by pauloxnet** [comment]
Ok but the generated secret key per se it's ok. If I start a projects and then move the generated secret key in my gitignored `.env` file everythin is secure. Maybe it's something to discuss in the ML.

**Comment by kosc** [comment]
Where is no "django-generated" in docs/spelling_wordlist. Should I add it or better just change to "generated by Django"?

File: docs/ref/checks.txt [hunk 7]
```diff
@@ -431,7 +431,7 @@ The following checks are run if you use the :option:`check --deploy` option:
   a load balancer or reverse-proxy server  to redirect all connections to HTTPS.
 * **security.W009**: Your :setting:`SECRET_KEY` has less than 50 characters or
   less than 50 unique characters, or it's a Django-generated secret key
-  starting with django-insecure. Please generate a long and random
+  starting with Django-insecure. Please generate a long and random
```
Comment (jacobtylerwalls) [inline docs/ref/checks.txt:434]: Hi there. I'm uncertain why this would be uppercased. Wouldn't it be better to use back ticks e.g. ``django-insecure``?

Comment (kosc) [reply docs/ref/checks.txt:434]: Hello!
Because there is no "django-generated" in docs/spelling_wordlist and I got a spelling error. I have no idea why uppercase fixed the error. I can just add new value to spelling_wordlist.

File: django/core/checks/security/base.py [hunk 8]
```diff
@@ -69,8 +69,9 @@
 
 W009 = Warning(
     "Your SECRET_KEY has less than %(min_length)s characters or less than "
-    "%(min_unique_chars)s unique characters. Please generate a long and random "
-    "SECRET_KEY, otherwise many of Django's security-critical features will be "
+    "%(min_unique_chars)s unique characters, or it's a Django-generated secret key starting with django-insecure."
+    "Please generate a sufficiently long and random SECRET_KEY, "
+    "otherwise many of Django's security-critical features will be "
     "vulnerable to attack." % {
```
Comment (pope1ni) [inline django/core/checks/security/base.py:75]: Things seem to be hung up on "Django-generated". I propose the following wording instead which seems clearer to me. I also note that "sufficiently" sneaked in there, but I'm not sure it really adds anything. Also note that the line wrapping should be maintained.

```suggestion
    "Your SECRET_KEY has less than %(min_length)s characters, less than "
    "%(min_unique_chars)s unique characters, or it is prefixed with "
    "'django-insecure-' indicating that it was generated automatically by "
    "Django. Please generate a long and random SECRET_KEY, otherwise many of "
    "Django's security-critical features will be vulnerable to attack." % {
```

(Note this suggestion may not apply cleanly in GitHub as I needed to include unchanged lines around the removed and added lines.)

File: docs/ref/checks.txt [hunk 9]
```diff
@@ -430,9 +430,10 @@ The following checks are run if you use the :option:`check --deploy` option:
   connections, you may want to either set this setting to ``True`` or configure
   a load balancer or reverse-proxy server  to redirect all connections to HTTPS.
 * **security.W009**: Your :setting:`SECRET_KEY` has less than 50 characters or
-  less than 5 unique characters. Please generate a long and random
-  ``SECRET_KEY``, otherwise many of Django's security-critical features will be
-  vulnerable to attack.
+  less than 50 unique characters, or it's a Django-generated secret key
+  starting with Django-insecure. Please generate a long and random
+  :setting:``SECRET_KEY``, otherwise many of Django's security-critical
+  features will be vulnerable to attack.
```
Comment (pope1ni) [inline docs/ref/checks.txt:436]: With the change to the wording of the message, the issue with "django-generated" should be gone. I've also added backticks around the string here:

```suggestion
* **security.W009**: Your :setting:`SECRET_KEY` has less than 50 characters,
  less than 5 unique characters, or it is prefixed with ``'django-insecure-'``
  indicating that it was generated automatically by Django. Please generate a
  long and random ``SECRET_KEY``, otherwise many of Django's security-critical
  features will be vulnerable to attack.
```

(Again, this suggestion may not apply cleanly in GitHub.)

File: tests/user_commands/tests.py [hunk 10]
```diff
@@ -333,7 +333,7 @@ def test_get_random_secret_key(self):
         key = get_random_secret_key()
         self.assertEqual(len(key), 50)
         for char in key:
-            self.assertIn(char, 'abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+)')
+            self.assertIn(char, 'abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+):')
```
Comment (pope1ni) [inline tests/user_commands/tests.py:336]: Please revert this change as we removed colons from the prefix string.

**Review by pope1ni — CHANGES_REQUESTED** [review]
Hi @kosc 

Thanks for your efforts on this. I've made some suggestions to improve the readability of the message.

I'm wondering if we should also introduce a constant for the prefix alongside the other constants?

```python
SECRET_KEY_INSECURE_PREFIX = 'django-insecure-'
```

File: docs/ref/checks.txt (hunk 7 above)
Comment (jacobtylerwalls) [reply docs/ref/checks.txt:434]: Ah, but I was referring to "Django-insecure", not "Django-generated." Nick's suggestion below adopts backticks as well as single-quotes for readability, which I think makes sense.

File: django/conf/global_settings.py [hunk 11]
```diff
@@ -261,6 +261,8 @@ def gettext_noop(s):
 #    ]
 IGNORABLE_404_URLS = []
 
+# Prefix for default SECRET_KEY (generated by Django)
+SECRET_KEY_INSECURE_PREFIX = 'django-insecure-'
```
Comment (pope1ni) [inline django/conf/global_settings.py:265]: I didn't mean adding this as a configurable setting as we don't want users to change it.
```suggestion
```

I meant adding it as a constant next to the existing constants related to checks on the secret key: https://github.com/django/django/blob/471d430f21a4cf6a51ecb4ae04a2c740c0771c41/django/core/checks/security/base.py#L12-L13

File: django/core/checks/security/base.py [hunk 12]
```diff
@@ -197,7 +197,7 @@ def check_secret_key(app_configs, **kwargs):
         passed_check = (
             len(set(secret_key)) >= SECRET_KEY_MIN_UNIQUE_CHARACTERS and
             len(secret_key) >= SECRET_KEY_MIN_LENGTH and
-            not settings.SECRET_KEY.startswith('django-insecure-')
+            not settings.SECRET_KEY.startswith(settings.SECRET_KEY_INSECURE_PREFIX)
```
Comment (pope1ni) [inline django/core/checks/security/base.py:200]:
```suggestion
            not settings.SECRET_KEY.startswith(SECRET_KEY_INSECURE_PREFIX)
```

File: tests/check_framework/test_security.py [hunk 13]
```diff
@@ -394,6 +394,10 @@ def test_missing_secret_key(self):
     def test_none_secret_key(self):
         self.assertEqual(base.check_secret_key(None), [base.W009])
 
+    @override_settings(SECRET_KEY='django-insecure-' + ('abcdefghijklmnopqrstuvwx' * 2))
```
Comment (pope1ni) [inline tests/check_framework/test_security.py:397]: We can use the constant here and the prefix also contains enough unique characters, so we can simplify:
```suggestion
    @override_settings(SECRET_KEY=base.SECRET_KEY_INSECURE_PREFIX + 'x' * 34)
```

File: docs/ref/checks.txt [hunk 14]
```diff
@@ -432,9 +432,10 @@ The following checks are run if you use the :option:`check --deploy` option:
   connections, you may want to either set this setting to ``True`` or configure
   a load balancer or reverse-proxy server  to redirect all connections to HTTPS.
 * **security.W009**: Your :setting:`SECRET_KEY` has less than 50 characters or
```
Comment (pope1ni) [inline docs/ref/checks.txt:434]:
```suggestion
* **security.W009**: Your :setting:`SECRET_KEY` has less than 50 characters,
```

File: django/core/checks/security/base.py [hunk 15]
```diff
@@ -69,9 +69,10 @@
 
 W009 = Warning(
     "Your SECRET_KEY has less than %(min_length)s characters or less than "
```
Comment (pope1ni) [inline django/core/checks/security/base.py:71]:
```suggestion
    "Your SECRET_KEY has less than %(min_length)s characters, less than "
```

File: django/core/checks/security/base.py [hunk 16]
```diff
@@ -69,9 +69,10 @@
...
-    "%(min_unique_chars)s unique characters. Please generate a long and random "
-    "SECRET_KEY, otherwise many of Django's security-critical features will be "
-    "vulnerable to attack." % {
+    "%(min_unique_chars)s unique characters, or it's prefixed with "
+    "'django-insecure-' indicating that it was generated automatically by "
+    "Django. Please generate a long and random SECRET_KEY, otherwise many of "
+    "Django's security-critical features will be vulnerable to attack." % {
```
Comment (pope1ni) [inline django/core/checks/security/base.py:75]:
```suggestion
    "%(min_unique_chars)s unique characters, or it's prefixed with %(prefix)s "
    "indicating that it was generated automatically by Django. Please "
    "generate a long and random SECRET_KEY, otherwise many of Django's "
    "security-critical features will be vulnerable to attack." % {
```

File: django/core/checks/security/base.py [hunk 17]
```diff
@@ -69,9 +69,10 @@
...
         'min_length': SECRET_KEY_MIN_LENGTH,
         'min_unique_chars': SECRET_KEY_MIN_UNIQUE_CHARACTERS,
```
Comment (pope1ni) [inline django/core/checks/security/base.py:77]:
```suggestion
        'min_unique_chars': SECRET_KEY_MIN_UNIQUE_CHARACTERS,
        'prefix': SECRET_KEY_INSECURE_PREFIX,
```

File: django/core/management/commands/startproject.py [hunk 18]
```diff
@@ -1,3 +1,4 @@
+from django.conf.global_settings import SECRET_KEY_INSECURE_PREFIX
```
Comment (pope1ni) [inline django/core/management/commands/startproject.py:1]:
```suggestion
from django.core.checks.security.base import SECRET_KEY_INSECURE_PREFIX
```

**Review by pope1ni — CHANGES_REQUESTED** [review]

**Review by pope1ni — APPROVED** [review]
Hi @kosc.

Thanks for the updates. This looks generally good to me now. I've removed the "patch needs improvement" flag in ticket-31757 so this should appear on the review queue again for the fellows.

File: tests/check_framework/test_security.py [hunk 19]
```diff
@@ -394,6 +394,10 @@ def test_missing_secret_key(self):
     def test_none_secret_key(self):
         self.assertEqual(base.check_secret_key(None), [base.W009])
 
+    @override_settings(SECRET_KEY=base.SECRET_KEY_INSECURE_PREFIX + 'x' * 34)
```
Comment (felixxm) [inline tests/check_framework/test_security.py:397]: We should use key that fails only on a prefix check.

**Review by felixxm — COMMENTED** [review]
@kosc Thanks :+1: Welcome aboard :boat: 

I've rebased and pushed small edits.
